File: assign2/scala/src/fibonacci/FibonacciSeries.scala
package fibonacci
import scala.annotation.tailrec

class FibonacciSeries {
	
	def fibonnaciRecursive(number : Int) : List[Int] = number match {
		case x if x < 0 => List()
		case 0 => List(1)
		case 1 => List(1, 1)
		case _ => {
			val series = fibonnaciRecursive(number-1)
			series :+ series.takeRight(2).sum
		}
	}
	
	@tailrec
	final def fibonnaciTailRecursive(number : Int, series : List[Int] = List(1)) : List[Int] = number match {
		case x if x < 0 => List()
		case 0 => series
		case _ => fibonnaciTailRecursive(number - 1, series :+ series.takeRight(2).sum)
	}
}

File: assign2/erlang/src/fibonacci_series.erl
-module(fibonacci_series).
-export([recursive_fibonacci/1, tail_recursive_fibonacci/1]).

recursive_fibonacci(0) -> [1];
recursive_fibonacci(1) -> [1, 1];
recursive_fibonacci(N) -> 
	Series = recursive_fibonacci(N - 1),
	Series ++ [lists:sum(lists:nthtail(N - 2, Series))].

tail_recursive_fibonacci(0) -> [1];
tail_recursive_fibonacci(N) -> tail_recursive_fibonacci(N, [1, 1]).

tail_recursive_fibonacci(1, Series) -> Series;
tail_recursive_fibonacci(N, Series) -> 
	tail_recursive_fibonacci(N - 1, Series ++ [lists:sum(lists:nthtail(length(Series) - 2, Series))]).



